From 6ac47e547b8cc0ec33ab272c88da43a13b4e0a67 Mon Sep 17 00:00:00 2001
From: Xing Zheng <zhengxing@rock-chips.com>
Date: Wed, 29 Aug 2018 20:51:46 +0800
Subject: [PATCH] ASoC: rk3308_codec: Fix the broken loopback sometimes

We need to insert some delay after enabling ADC current
and waiting ADCs are stable for BIST mode mainly.

Change-Id: Ib3cdc6aa36f8674ba8d8defadb47baac72f4745e
Signed-off-by: Xing Zheng <zhengxing@rock-chips.com>
---
 sound/soc/codecs/rk3308_codec.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/sound/soc/codecs/rk3308_codec.c b/sound/soc/codecs/rk3308_codec.c
index 15b85d1d5dc1..ae530d341242 100644
--- a/sound/soc/codecs/rk3308_codec.c
+++ b/sound/soc/codecs/rk3308_codec.c
@@ -2366,6 +2366,14 @@ static int rk3308_codec_adc_ana_enable(struct rk3308_codec_priv *rk3308,
 				   RK3308_ADC_CURRENT_EN);
 	}
 
+	/*
+	 * This is mainly used for BIST mode that wait ADCs are stable.
+	 *
+	 * By tested results, the type delay is >40us, but we need to leave
+	 * enough delay margin.
+	 */
+	usleep_range(400, 500);
+
 	/* vendor step 4*/
 	for (idx = 0; adc_for_each_grp(rk3308, type, idx, &grp); idx++) {
 		if (grp < 0 || grp > ADC_LR_GROUP_MAX - 1)
-- 
2.35.3

